### PR TITLE
Issue #3185523 by rolki: Implement basic topic querying in the Graph

### PR DESCRIPTION
**Pull request by rollki**
## Problem
With the intention to allow external websites to display teasers for topics. Those teasers should have access to the author, title, created data, topic type, image, contents and comment data. Additionally it should be able to provide a link to the topic so that it can be viewed on the Open Social website.

The updated date is currently omitted because this information is not generally displayed on an Open Social platform itself.

It should be possible to

- Select a topic by its UUID
- Select an overview of all topics and paginate through results according to the Relay connection specification

Querying and searching for data should respect existing privacy measures (such as visibility settings) according to platform configuration.

## Solution
Implement the required GraphQL Schema and DataProducer plugins in `social_topic` that extend the schema defined on `social_graphql`

## Issue tracker
- https://www.drupal.org/project/social/issues/3185523
- https://getopensocial.atlassian.net/browse/YANG-5684
- https://github.com/drupal-graphql/graphql/issues/1219

## How to test
- [ ] Enable `social_graphql` module
- [ ] Apply https://github.com/goalgorilla/open_social/pull/2400 as a patch
- [ ] Use `Altair` or some similar tool to send a request
- [ ] Create some events on the platform
- [ ] Send request, like:
```
  topics(first: 2) {
    nodes {
      id
      title
    }
  }
```

### CODE REVIEW

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2343-jppkwgejhfrrlphogdklmaxp5oa2suj8.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/60d32c8af3ee306f29f6ee85

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2343-jppkwgejhfrrlphogdklmaxp5oa2suj8.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/60d32c8af3ee306f29f6ee85

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2343-jppkwgejhfrrlphogdklmaxp5oa2suj8.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/60d32c8af3ee306f29f6ee85

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2343-jppkwgejhfrrlphogdklmaxp5oa2suj8.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/60d32c8af3ee306f29f6ee85

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2343-jppkwgejhfrrlphogdklmaxp5oa2suj8.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/60d32c8af3ee306f29f6ee85

File: modules/social_features/social_topic/tests/src/Kernel/GraphQL/QueryTopicsTest.php [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+    // The default topic config contains a body text field.
+    'field',
+    'text',
+    'filter',
+    'file',
+    'image',
+    // For the comment functionality.
+    'social_comment',
+    'comment',
+    'menu_ui',
```
Comment (rollki) [inline modules/social_features/social_topic/tests/src/Kernel/GraphQL/QueryTopicsTest.php:43]: Check configuration.

File: modules/social_features/social_topic/tests/src/Kernel/GraphQL/QueryTopicsTest.php [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+  public function testAnonymousUserCanNotViewUnpublishedTopics() {
+    $this->createNode([
+      'type' => 'topic',
+      'field_content_visibility' => 'public',
+      'status' => NodeInterface::NOT_PUBLISHED,
+    ]);
+    $this->createNode([
+      'type' => 'topic',
+      'field_content_visibility' => 'public',
+      'status' => NodeInterface::NOT_PUBLISHED,
```
Comment (rollki) [inline modules/social_features/social_topic/tests/src/Kernel/GraphQL/QueryTopicsTest.php:172]: make published

File: modules/social_features/social_topic/tests/src/Kernel/GraphQL/QueryTopicsTest.php [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+      static fn($topic) => $topic->uuid(),
+      $topics ?? []
+    );
+
+    $this->assertEndpointSupportsPagination(
+      'topics',
+      $topic_uuids
+    );
+  }
+
```
Comment (rollki) [inline modules/social_features/social_topic/tests/src/Kernel/GraphQL/QueryTopicsTest.php:105]: Add a test if the user doesn't have permission to see public topics

File: modules/social_features/social_topic/tests/src/Kernel/GraphQL/QueryTopicsTest.php (hunk 1 above)
Comment (rollki) [reply modules/social_features/social_topic/tests/src/Kernel/GraphQL/QueryTopicsTest.php:43]: There is dependencies in `node.type.topic.yml` file, so not sure if we can scrap it
dependencies:
  module:
    - menu_ui

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2343-jppkwgejhfrrlphogdklmaxp5oa2suj8.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/60d32c8af3ee306f29f6ee85

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2343-jppkwgejhfrrlphogdklmaxp5oa2suj8.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/60d32c8af3ee306f29f6ee85

File: modules/social_features/social_comment/graphql/social_comment_schema_extension.base.graphqls [hunk 4]
```diff
@@ -1,3 +1,23 @@
+"""
+Commentable interface that contains fields that present in the commentable entity.
+"""
+interface Commentable {
```
Comment (rollki) [inline modules/social_features/social_comment/graphql/social_comment_schema_extension.base.graphqls:4]: `Commentable` interface added to `social_comment` module, so not sure if there's any reason to create another PR for this since we already have a lot of PR's :|

File: modules/social_features/social_comment/src/Plugin/GraphQL/DataProducer/QueryComments.php [hunk 5]
```diff
@@ -77,13 +83,13 @@ class QueryComments extends EntityDataProducerPluginBase {
    * @return \Drupal\social_graphql\GraphQL\ConnectionInterface
    *   An entity connection with results and data about the paginated results.
    */
-  public function resolve(?NodeInterface $parent, ?int $first, ?string $after, ?int $last, ?string $before, bool $reverse, string $sortKey, RefinableCacheableDependencyInterface $metadata) {
-    if (is_string($parent)) {
```
Comment (rollki) [inline modules/social_features/social_comment/src/Plugin/GraphQL/DataProducer/QueryComments.php:81]: It was not correct condition 😄

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2343-jppkwgejhfrrlphogdklmaxp5oa2suj8.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/60d32c8af3ee306f29f6ee85

File: modules/social_features/social_comment/src/Plugin/GraphQL/DataProducer/QueryComments.php [hunk 6]
```diff
@@ -77,13 +83,13 @@ class QueryComments extends EntityDataProducerPluginBase {
...
-      $nodes = $this->entityTypeManager->getStorage('node')->loadByProperties(['uuid' => $parent]);
+  public function resolve(?NodeInterface $parent, ?string $bundle, ?int $first, ?string $after, ?int $last, ?string $before, bool $reverse, string $sortKey, RefinableCacheableDependencyInterface $metadata) {
+    if ($parent) {
+      $nodes = $this->entityTypeManager->getStorage('node')->loadByProperties(['uuid' => $parent->uuid()]);
       $parent = reset($nodes);
     }
 
-    $query_helper = new CommentQueryHelper($sortKey, $this->entityTypeManager, $this->graphqlEntityBuffer, $parent);
+    $query_helper = new CommentQueryHelper($sortKey, $this->entityTypeManager, $this->graphqlEntityBuffer, $parent, $bundle);
```
Comment (rollki) [inline modules/social_features/social_comment/src/Plugin/GraphQL/DataProducer/QueryComments.php:92]: We added a `bundle` since produced query can get comments from different types, so it was possible to see an incorrect comment list on the topic (e.g. we were able to see a comment from a group post as a topic comment in GraphQL response).

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2343-jppkwgejhfrrlphogdklmaxp5oa2suj8.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/60d32c8af3ee306f29f6ee85

**Comment by rollki** [comment]
So, this PR should be merged after https://github.com/goalgorilla/open_social/pull/2400

**Comment by Kingdutch** [comment]
@rollki Can we use the following configuration available in `KernelTestBase` (can be used in any Kernel test) to ignore the failing schema's in these tests.

As long as each ignored configuration item has a `// @todo` that points to a specific Drupal.org issue then we know what's needed to improve the code and this doesn't need to wait on #2400 : ) (We shouldn't ignore config just to make the test pass, but if it's linked to an issue that describes the underlying problem then it's a tool we can use)

```
  /**
   * An array of config object names that are excluded from schema checking.
   *
   * @var string[]
   */
  protected static $configSchemaCheckerExclusions = [
  ];  
```

File: modules/social_features/social_comment/graphql/social_comment_schema_extension.base.graphqls [hunk 7]
```diff
@@ -1,3 +1,23 @@
+"""
+Commentable interface that contains fields that present in the commentable entity.
```
Comment (Kingdutch) [inline modules/social_features/social_comment/graphql/social_comment_schema_extension.base.graphqls:2]:
```suggestion
Things that can be commented on.
```

File: modules/social_features/social_topic/graphql/social_topic_schema_extension.base.graphqls [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+}
+
+"""
+The set of valid sort keys for the topic query.
+"""
+enum TopicSortKeys {
+  """Sort by topic creation date"""
+  CREATED_AT
+}
+
```
Comment (Kingdutch) [inline modules/social_features/social_topic/graphql/social_topic_schema_extension.base.graphqls:84]:
```suggestion

"""
The type of a topic.

Each topic is tagged with a topic type to make it clear what this topic is about.
A topic can be a piece of news, a blogpost, a document, or another type depending on the platform.
"""
```

File: modules/social_features/social_topic/graphql/social_topic_schema_extension.base.graphqls [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+type TopicType {
+  id: ID!
```
Comment (Kingdutch) [inline modules/social_features/social_topic/graphql/social_topic_schema_extension.base.graphqls:86]:
```suggestion
  """
  The unique identifier for the topic type.
  """
  id: ID!
```

File: modules/social_features/social_topic/graphql/social_topic_schema_extension.base.graphqls [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+  label: String
```
Comment (Kingdutch) [inline modules/social_features/social_topic/graphql/social_topic_schema_extension.base.graphqls:87]:
```suggestion
  """
  The human readable name of the topic type.
  """
  label: String
```

File: modules/social_features/social_topic/src/Plugin/GraphQL/QueryHelper/TopicQueryByTypeHelper.php [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+      ->getQuery()
+      ->condition('uuid', $this->type)
+      ->execute();
+
+    return $this->entityTypeManager->getStorage('node')
+      ->getQuery()
+      ->currentRevision()
+      ->accessCheck(TRUE)
+      ->condition('type', 'topic')
+      ->condition('field_topic_type', $term_ids ?: [0], 'IN');
```
Comment (Kingdutch) [inline modules/social_features/social_topic/src/Plugin/GraphQL/QueryHelper/TopicQueryByTypeHelper.php:58]: The following should work as a single query :)

```suggestion
    return $this->entityTypeManager->getStorage('node')
      ->getQuery()
      ->currentRevision()
      ->accessCheck(TRUE)
      ->condition('type', 'topic')
      ->condition('field_topic_type.0.entity:taxonomy_term.uuid', $this->type);
```

For other places where you may need to load an entity by `uuid` you can do so with `$this->entityRepository->loadEntityByUuid('entity_type', $uuid)` :D That avoids scenarios in which the `uuid` is not actually in the `uuid` field (which I haven't seen done, but is technically possible 🙈 )

File: modules/social_features/social_topic/tests/src/Kernel/GraphQL/QueryTopicsTest.php (hunk 1 above)
Comment (Kingdutch) [reply modules/social_features/social_topic/tests/src/Kernel/GraphQL/QueryTopicsTest.php:43]:
```suggestion
    // node.type.topic has a configuration dependency on the menu_ui module.
    'menu_ui',
```

**Review by Kingdutch — COMMENTED** [review]
Looking good! One last query optimisation and some polishing of comments and we're good to go!

Let me know if you have time to make the change or want me to do this before merging.

File: modules/social_features/social_topic/src/Plugin/GraphQL/SchemaExtension/TopicSchemaExtension.php [hunk 12]
```diff
@@ -0,0 +1,164 @@
...
+      )
+    );
+
+    $registry->addFieldResolver('Topic', 'author',
+      $builder->produce('entity_owner')
+        ->map('entity', $builder->fromParent())
+    );
+
+    $registry->addFieldResolver('Topic', 'bodyHtml',
+      $builder->fromPath('entity:node', 'body.0.value')
```
Comment (Kingdutch) [inline modules/social_features/social_topic/src/Plugin/GraphQL/SchemaExtension/TopicSchemaExtension.php:60]: This needs rendering. See https://github.com/goalgorilla/open_social/blob/55e5b28c2750bd9d4a5da1b18c4c4836fed67a65/modules/social_features/social_comment/src/Plugin/GraphQL/SchemaExtension/CommentSchemaExtension.php#L50

The field value may contain Drupal specific tags that are not valid HTML.

File: modules/social_features/social_topic/src/Plugin/GraphQL/SchemaExtension/TopicSchemaExtension.php [hunk 13]
```diff
@@ -0,0 +1,164 @@
...
+      $builder->produce('entity_uuid')
+        ->map('entity', $builder->fromParent())
+    );
+
+    $registry->addFieldResolver('Topic', 'title',
+      $builder->compose(
+        $builder->produce('entity_label')
+          ->map('entity', $builder->fromParent()),
+        $builder->produce('uppercase')
+          ->map('string', $builder->fromParent())
```
Comment (Kingdutch) [inline modules/social_features/social_topic/src/Plugin/GraphQL/SchemaExtension/TopicSchemaExtension.php:50]: Why are we uppercasing the label here?

File: modules/social_features/social_topic/src/Plugin/GraphQL/SchemaExtension/TopicSchemaExtension.php [hunk 14]
```diff
@@ -0,0 +1,164 @@
...
+    );
+
+    $registry->addFieldResolver('Topic', 'heroImage',
+      $builder->produce('field')
+        ->map('entity', $builder->fromParent())
+        ->map('field', $builder->fromValue('field_topic_image'))
+    );
+
+    $registry->addFieldResolver('Topic', 'created',
+      $builder->fromPath('entity:node', 'created.value')
```
Comment (Kingdutch) [inline modules/social_features/social_topic/src/Plugin/GraphQL/SchemaExtension/TopicSchemaExtension.php:70]:
```suggestion
      $builder->product('entity_created')
        ->map('entity', $builder->fromParent())
```

File: modules/social_features/social_topic/src/Plugin/GraphQL/SchemaExtension/TopicSchemaExtension.php [hunk 15]
```diff
@@ -0,0 +1,164 @@
...
+        ->map('sortKey', $builder->fromArgument('sortKey'))
+    );
+
+    $registry->addFieldResolver('Topic', 'url',
+      $builder->compose(
+        $builder->produce('entity_url')
+          ->map('entity', $builder->fromParent()),
+        $builder->produce('url_path')
+          ->map('url', $builder->fromParent())
+      )
```
Comment (Kingdutch) [inline modules/social_features/social_topic/src/Plugin/GraphQL/SchemaExtension/TopicSchemaExtension.php:91]: I think `url_path` produces only the /path right? I think an absolute URL is expected here.

File: modules/social_features/social_topic/src/Plugin/GraphQL/SchemaExtension/TopicSchemaExtension.php [hunk 16]
```diff
@@ -0,0 +1,164 @@
...
+    );
+
+    $registry->addFieldResolver('Topic', 'created',
+      $builder->fromPath('entity:node', 'created.value')
+    );
```
Comment (Kingdutch) [inline modules/social_features/social_topic/src/Plugin/GraphQL/SchemaExtension/TopicSchemaExtension.php:96]:
```suggestion
```

This is duplicated from line 69

**Comment by Kingdutch** [comment]
Sorry for all the separate messages. I caught some things while looking at the Event Schema class that were also present here.

My review comments also reveal some lacking test coverage. We should create at least a single test case that queries all fields on the type. This can be used to test that the `topic` field (query by ID) works.

File: modules/social_features/social_topic/src/Plugin/GraphQL/SchemaExtension/TopicSchemaExtension.php (hunk 13 above)
Comment (rollki) [reply modules/social_features/social_topic/src/Plugin/GraphQL/SchemaExtension/TopicSchemaExtension.php:50]: I don't know, we did it everywhere if I'm not mistaken, but I'm also not sure if it is a good idea. Probably because of `composable-schemas.md` - html/modules/contrib/graphql/doc/advanced/composable-schemas.md

File: modules/social_features/social_topic/src/Plugin/GraphQL/SchemaExtension/TopicSchemaExtension.php (hunk 12 above)
Comment (rollki) [reply modules/social_features/social_topic/src/Plugin/GraphQL/SchemaExtension/TopicSchemaExtension.php:60]: Yeah, but it will return a lot of HTML, that is probably not needed,

E.g. with how it is implemented in comments:
Снимок экрана 2021-07-06 в 17 14 27

E.g. with `$builder->fromPath('entity:node', 'body.0.value')`:
Снимок экрана 2021-07-06 в 17 15 25

But if you think the first example is correct then I'll change it.

File: modules/social_features/social_topic/src/Plugin/GraphQL/SchemaExtension/TopicSchemaExtension.php (hunk 14 above)
Comment (rollki) [reply modules/social_features/social_topic/src/Plugin/GraphQL/SchemaExtension/TopicSchemaExtension.php:70]:
```
$builder->produce('entity_created')
        ->map('entity', $builder->fromParent())
        ->map('format', $builder->fromValue('U'))
```

File: modules/social_features/social_topic/src/Plugin/GraphQL/SchemaExtension/TopicSchemaExtension.php (hunk 15 above)
Comment (rollki) [reply modules/social_features/social_topic/src/Plugin/GraphQL/SchemaExtension/TopicSchemaExtension.php:91]: Yeah, but we should create a custom Data Producer here I guess :(

File: modules/social_features/social_comment/graphql/social_comment_schema_extension.base.graphqls (hunk 7 above)
Comment (rollki) [reply modules/social_features/social_comment/graphql/social_comment_schema_extension.base.graphqls]: Changed.

File: modules/social_features/social_topic/graphql/social_topic_schema_extension.base.graphqls (hunk 8 above)
Comment (rollki) [reply modules/social_features/social_topic/graphql/social_topic_schema_extension.base.graphqls:84]: Changed.

File: modules/social_features/social_topic/graphql/social_topic_schema_extension.base.graphqls (hunk 9 above)
Comment (rollki) [reply modules/social_features/social_topic/graphql/social_topic_schema_extension.base.graphqls:95]: Added

File: modules/social_features/social_topic/src/Plugin/GraphQL/QueryHelper/TopicQueryByTypeHelper.php (hunk 11 above)
Comment (rollki) [reply modules/social_features/social_topic/src/Plugin/GraphQL/QueryHelper/TopicQueryByTypeHelper.php]: Yeah, good to know 😁

File: modules/social_features/social_topic/src/Plugin/GraphQL/SchemaExtension/TopicSchemaExtension.php (hunk 13 above)
Comment (rollki) [reply modules/social_features/social_topic/src/Plugin/GraphQL/SchemaExtension/TopicSchemaExtension.php]: Removed uppercase.

File: modules/social_features/social_topic/src/Plugin/GraphQL/SchemaExtension/TopicSchemaExtension.php (hunk 14 above)
Comment (rollki) [reply modules/social_features/social_topic/src/Plugin/GraphQL/SchemaExtension/TopicSchemaExtension.php]: Changed.

File: modules/social_features/social_topic/src/Plugin/GraphQL/SchemaExtension/TopicSchemaExtension.php (hunk 16 above)
Comment (rollki) [reply modules/social_features/social_topic/src/Plugin/GraphQL/SchemaExtension/TopicSchemaExtension.php]: Removed.

File: modules/social_features/social_topic/tests/src/Kernel/GraphQL/QueryTopicsTest.php (hunk 1 above)
Comment (rollki) [reply modules/social_features/social_topic/tests/src/Kernel/GraphQL/QueryTopicsTest.php:44]: Added

File: modules/social_features/social_topic/src/Plugin/GraphQL/SchemaExtension/TopicSchemaExtension.php (hunk 12 above)
Comment (Kingdutch) [reply modules/social_features/social_topic/src/Plugin/GraphQL/SchemaExtension/TopicSchemaExtension.php:56]: The `path` version will break as soon as you add an image though. So in my opinion that's not an option.

If the excess HTML wrapping is an issue when rendering the field, then we should figure out a way to properly configure fields for rendering in GraphQL (e.g. without label and without extra field wrappers). That probably means we'll need to add extra view modes for all content and fields though, so I think that's outside of the scope for this PR and an improvement we can make in the future.

The extra included div is slightly unfortunate in that it uses quite generic classes but that's something that people should be able to work around relatively easily for the time being.

File: modules/social_features/social_topic/src/Plugin/GraphQL/SchemaExtension/TopicSchemaExtension.php (hunk 15 above)
Comment (rollki) [reply modules/social_features/social_topic/src/Plugin/GraphQL/SchemaExtension/TopicSchemaExtension.php:89]: or we can do something like
```
    $registry->addFieldResolver('Topic', 'url',
      $builder->callback(function ($parent, $args) {
        return $parent->toUrl('canonical', ['absolute' => TRUE])->toString();
      })
    );
```

Comment (Kingdutch) [reply modules/social_features/social_topic/src/Plugin/GraphQL/SchemaExtension/TopicSchemaExtension.php:89]: There's nothing wrong with a custom data producer if it solves our problem and other data producers don't do that.

We can also take a closer look at the data producers we already have. Then we see `UrlPath` actually just turns a URL object into a string. So whether the returned URL is absolute depends on the configuration of the URL object passed to it.

Then we look at where that object comes from which is `EntityUrl` and we see it only does `$entity->toUrl()`. However, `EntityInterface::toUrl` does take options to that allow us to make the URL absolute. 

So rather than creating an entirely new DataProducer we can also make a contribution to the GraphQL module to provide optional `rel` and `options` arguments to the `entity_url` data producer. After we propose that as a PR to the GraphQL module we can add it to Open Social in two ways until it's merged.

1. We can make a patch and apply it to Open Social. This breaks as soon as the next minor version of the GraphQL module is released.
2. We can copy our PR into `social_core` as the `EntityUrl` class and give it the `entity_url_with_options` ID. Then add a comment to that class marking it as deprecated pointing to the PR in the GraphQL module. Once the GraphQL module's version is merged, we can update the code to use that class instead and then remove our own copy.

2 is only a little bit more work up-front, but is much less likely to break in the next weeks/months. So I think a contribution to the GraphQL module with option 2 to bring it into the distribution is desirable.

File: modules/social_features/social_topic/src/Plugin/GraphQL/SchemaExtension/TopicSchemaExtension.php (hunk 12 above)
Comment (rollki) [reply modules/social_features/social_topic/src/Plugin/GraphQL/SchemaExtension/TopicSchemaExtension.php]: Implemented as in `comments` for now.

File: modules/social_features/social_topic/src/Plugin/GraphQL/SchemaExtension/TopicSchemaExtension.php (hunk 15 above)
Comment (rollki) [reply modules/social_features/social_topic/src/Plugin/GraphQL/SchemaExtension/TopicSchemaExtension.php]: So, I chose option 2 :) 
I have created PR to the `graphql` module, which provides the ability to get an absolute URL.

See https://github.com/drupal-graphql/graphql/issues/1219

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2343-jppkwgejhfrrlphogdklmaxp5oa2suj8.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/60d32c8af3ee306f29f6ee85

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2343-jppkwgejhfrrlphogdklmaxp5oa2suj8.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/60d32c8af3ee306f29f6ee85

**Comment by rollki** [comment]
Hey @Kingdutch 
I have updated PR, with the following changes:

- I used `field_renderer` to view `bodyHtml` for `topics`, see https://github.com/goalgorilla/open_social/pull/2343/commits/bda2f16fdc41ef33132e3b3a1ecb4788f8d7f6e6#diff-f6de53b2570385a971a77b51d006ef7fb5ca5755e47ff83962b8a6eb8a95ebb2R56-R62 
- Created PR to the `drupal-graphql` module where was added the ability to view absolute entity URL, see https://github.com/drupal-graphql/graphql/pull/1220
- The last thing is your comment about testing all fields (https://github.com/goalgorilla/open_social/pull/2343#issuecomment-874645634), so I have created a separate test to resolve this, but there's only one method that queries all available fields, not sure if we also should test permissions there since it is already tested in `QueryTopicsTest.php`. See https://github.com/goalgorilla/open_social/blob/bda2f16fdc41ef33132e3b3a1ecb4788f8d7f6e6/modules/social_features/social_topic/tests/src/Kernel/GraphQL/QueryTopicTest.php

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2343-jppkwgejhfrrlphogdklmaxp5oa2suj8.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/60d32c8af3ee306f29f6ee85

File: composer.json [hunk 17]
```diff
@@ -106,6 +106,9 @@
             },
             "drupal/views_infinite_scroll" : {
                 "Headers in table format repeat on load more instead of adding rows": "https://www.drupal.org/files/issues/2019-08-15/table_tbody_append-2899705-26.patch"
+            },
+            "drupal/graphql": {
+                "Add the ability to get absolute URL of entity": "https://patch-diff.githubusercontent.com/raw/drupal-graphql/graphql/pull/1220.diff"
```
Comment (Kingdutch) [inline composer.json:111]: Great work on making the contribution to the GraphQL module here! However, while that's still under review, lets create a copy of the Data Producer in Open Social with a different ID and a `@deprecated Remove when #PR is merged`. This allows us to use it in Open Social already, but ensures we don't get patch conflicts when GraphQL 4.2 is removed. Then we can clean up our own code when we upgrade to 4.2.

**Review by Kingdutch — CHANGES_REQUESTED** [review]
Great work Roman! 

> The last thing is your comment about testing all fields (Issue #3185523 by rolki: Implement basic topic querying in the Graph #2343 (comment)), so I have created a separate test to resolve this, but there's only one method that queries all available fields, not sure if we also should test permissions there since it is already tested in QueryTopicsTest.php. See https://github.com/goalgorilla/open_social/blob/bda2f16fdc41ef33132e3b3a1ecb4788f8d7f6e6/modules/social_features/social_topic/tests/src/Kernel/GraphQL/QueryTopicTest.php

I think the test that makes sure all fields are present with the correct output is fine! :) The only test coverage we should add is a test that tries to make a query without the permissions and that it provides a NULL or access denied response as needed.

That way we make sure that we're actually applying the access checks :D

**Comment by rollki** [comment]
@Kingdutch 
An additional test was added to test querying without permissions, data producer for entity URL was also created and marked as deprecated.
See https://github.com/goalgorilla/open_social/pull/2343/commits/04bfc2debfb1d501c970f6d8049486d1a70c0547

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2343-jppkwgejhfrrlphogdklmaxp5oa2suj8.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/60d32c8af3ee306f29f6ee85

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2343-jppkwgejhfrrlphogdklmaxp5oa2suj8.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/60d32c8af3ee306f29f6ee85

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2343-jppkwgejhfrrlphogdklmaxp5oa2suj8.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/60d32c8af3ee306f29f6ee85

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2343-jppkwgejhfrrlphogdklmaxp5oa2suj8.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/60d32c8af3ee306f29f6ee85

**Comment by Kingdutch** [comment]
Squashed all commits in a rebase since the existing history didn't allow me to split it up in easier chunks. Also moved the custom `EntityUrl` DataProducer to `social_graphql` since we'll need it in multiple places (and the ID makes it path agnostic).

This also ensures we don't accidentally merge 10.2.x changes back into 10.3.x (since this PR was started on 10.2.x and swapped to 10.3.x, which is dangerous).

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2343-jppkwgejhfrrlphogdklmaxp5oa2suj8.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/60d32c8af3ee306f29f6ee85

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2343-jppkwgejhfrrlphogdklmaxp5oa2suj8.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/60d32c8af3ee306f29f6ee85

**Comment by Kingdutch** [comment]
Merged after discussing with Nucleus. The Travis test failures are caused by another merge to 10.3.x and this PR does not contain code that affects Travis tests. PHPUnit covers the code in the PR and is green.
